src/library/LoansWithSettlement.sol

```diff
@@ -0,0 +1,198 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8;
+
+import {IFlashLoanSolverWrapper} from "../interface/IFlashLoanSolverWrapper.sol";
+import {IERC20} from "../vendored/IERC20.sol";
+
+library BytesUtil {
+    uint256 private constant BYTES_CONTENT_OFFSET = 32;
+
+    /// @dev Allocate a bytes array in memory with arbitrary data in it.
+    /// This is cheaper than `new bytes(length)` because it doesn't zero the
+    /// content of the array. It is supposed to be used when the newly allocated
+    /// memory will be fully overwritten at a later step.
+    function allocate(uint256 length) internal pure returns (bytes memory array) {
+        // https://docs.soliditylang.org/en/v0.8.26/internals/layout_in_memory.html
+        assembly ("memory-safe") {
+            array := mload(0x40)
+            mstore(array, length)
+            mstore(0x40, add(add(array, BYTES_CONTENT_OFFSET), length))
+        }
+    }
+
+    function memoryPointerToArray(bytes memory array) internal pure returns (uint256 ref) {
+        assembly ("memory-safe") {
+            // The fist word is the overall length of the memory vector, the
+            // second is the start of the content.
+            ref := array
+        }
+    }
+
+    function memoryPointerToContent(bytes memory array) internal pure returns (uint256 ref) {
+        // Arrays allocated by Solidity cannot cause an overflow, since a
+        // transaction would run out of gas long before reaching the length
+        // needed for an overflow. Arrays that were manually allocated through
+        // assembly may cause an overflow, but any attempt to read or write to
+        // them would cause an out-of-gas revert.
+        unchecked {
+            ref = memoryPointerToArray(array) + BYTES_CONTENT_OFFSET;
+        }
+    }
+}
+
+library LoanRequest {
+    struct Data {
+        /// @notice The amount requested from the lender.
+        uint256 amount;
+        IFlashLoanSolverWrapper borrower;
+        address lender;
```

Comment:
Not sure we actually need these two fields. If we have one wrapper per lender, I'd assume the lender be part of the borrower's state.